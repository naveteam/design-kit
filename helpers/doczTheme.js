import React from 'react'

import { Provider } from 'src/Theme'

export default ({ children }) => {
  return <Provider>{children}</Provider>
}
