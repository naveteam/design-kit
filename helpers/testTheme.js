import React from 'react'
import { render } from '@testing-library/react'

import { Provider } from 'src/Theme'

export function renderWithTheme(component) {
  return render(<Provider>{component}</Provider>)
}
