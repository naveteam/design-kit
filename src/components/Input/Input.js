import React from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'
import { Typography } from 'src'
import modifiers from 'src/helpers/modifiers'

const Input = props => {
  const { label, ...rest } = props

  return (
    <Wrapper {...rest}>
      {label && <Label {...rest}>{label}</Label>}
      <InputContainer {...rest}>
        {rest.leftIcon && rest.leftIcon}
        <StyledInput disabled={rest.disabled} {...rest} />
        {rest.rightIcon && rest.rightIcon}
      </InputContainer>
      {rest.error && <Error>{rest.error}</Error>}
      {!rest.error && rest.caption && <Caption>{rest.caption}</Caption>}
    </Wrapper>
  )
}

const Wrapper = styled.div`
  ${props => modifiers(props)}
  ${({ theme: { colors }, disabled, error, verified }) => {
    if (disabled || error || verified) {
      return ''
    }
    return `
      &:focus-within {
        div {
          border: 1px solid ${colors.blue};
        }
        p:first-child {
          color: ${colors.blue};
        }
      }
    `
  }}
`

const InputContainer = styled.div`
  box-sizing: border-box;
  border-radius: 4px;
  transition: all 0.3s linear;
  display: flex;
  align-items: center;
  width: 328px;
  height: 36px;
  padding: 8px 10px;
  border: 1px solid ${({ theme }) => theme.colors.mediumGrey};
  ${({ theme: { colors }, ...props }) => {
    if (props.disabled) {
      return `
        border: 1px solid rgba(37, 37, 37, 0.3);
        background: ${colors.grey};`
    }
    if (props.error) {
      return `border: 1px solid ${colors.red};`
    }
    if (props.verified) {
      return `border: 1px solid ${colors.green};`
    }
  }}
`

const Label = styled(Typography)`
  margin: 0;
  margin-bottom: 5px;
  transition: all 0.3s linear;
  ${({ theme: { colors }, ...props }) => {
    if (props.disabled) {
      return `color: ${colors.mediumGrey};`
    }
    if (props.error) {
      return `color: ${colors.red};`
    }
  }}
`

const StyledInput = styled.input`
  font-family: Roboto;
  font-style: normal;
  font-weight: normal;
  font-size: 16px;
  line-height: 19px;
  border: none;
  outline: none;
  flex: 1;
  height: 100%;
  background: transparent;
  padding-left: ${props => (props.leftIcon ? '4px' : 0)};
  padding-right: ${props => (props.rightIcon ? '4px' : 0)};
  color: ${props => props.theme.colors.darkGrey};
  ::placeholder {
    color: rgba(37, 37, 37, 0.3);
  }
`

const Error = styled(Typography)`
  margin: 0;
  margin-top: 5px;
  color: ${({ theme }) => theme.colors.red};
  font-size: 12px;
`

const Caption = styled(Typography)`
  margin: 0;
  margin-top: 5px;
  color: ${({ theme }) => theme.colors.mediumGrey};
  font-size: 12px;
`

Input.propTypes = {
  label: PropTypes.string,
  error: PropTypes.string,
  caption: PropTypes.string,
  leftIcon: PropTypes.element,
  rightIcon: PropTypes.element,
  verified: PropTypes.bool
}

export default Input
