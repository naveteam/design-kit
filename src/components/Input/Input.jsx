import React, { useState, forwardRef } from 'react'
import styled, { css } from '@xstyled/styled-components'
import { variant } from '@xstyled/system'

import { Flex, Box } from '../Grid'
import { Typography } from '../'

const Input = forwardRef(({ label, message, prefix, suffix, error, placeholder, width }, ref) => {
  const [focus, setFocus] = useState(false)
  return (
    <Wrapper error={error} width={width}>
      <Label>{label}</Label>
      <Container focus={focus}>
        {prefix && <Affix forwardedAs='span'>{prefix}</Affix>}
        <InputBase ref={ref} placeholder={placeholder} onFocus={() => setFocus(true)} onBlur={() => setFocus(false)} />
        {suffix && <Affix forwardedAs='span'>{suffix}</Affix>}
      </Container>
      <Message>{message}</Message>
    </Wrapper>
  )
})

const errorVariant = variant({
  prop: 'error',
  default: false,
  variants: {
    true: css`
      p {
        color: red.400;
      }
      div {
        border-color: red.400;
        span {
          color: red.400;
        }
      }
    `
  }
})

const focusVariant = variant({
  prop: 'focus',
  default: false,
  variants: {
    true: css`
      border-color: blue.50;
      border-width: 2px;
      padding: 3px;
    `
  }
})

const Wrapper = styled(Box)`
  ${errorVariant}
`
const Label = styled(Typography)`
  font-size: 2;
  line-height: 1;
  font-weight: 1;
  margin-bottom: 2;
`
const Container = styled(Flex)`
  border-width: 1px;
  border-style: solid;
  border-color: black;
  border-radius: 2;
  padding: 2;
  align-items: center;
  ${focusVariant}
`
const Affix = styled(Typography)`
  font-size: 1;
  line-height: 1;
  margin: 0 2;
  pointer-events: none;
`
const InputBase = styled.input`
  border: 0;
  flex: 1;
  font-size: 3;
  line-height: 3;
  &::placeholder {
    color: gray.600;
  }
`
const Message = styled(Typography)`
  font-size: 1;
  line-height: 1;
  margin-top: 2;
`

export default Input
