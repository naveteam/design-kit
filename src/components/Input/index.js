import Input from './Input'

export * from './Input'

export { Input }
