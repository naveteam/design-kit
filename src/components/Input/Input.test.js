import React from 'react'

import { renderWithTheme } from 'helpers/testTheme'

import Input from './Input'

describe('Input', () => {
  it('renders input with label', () => {
    const { container } = renderWithTheme(<Input label='Label' />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('renders input with caption', () => {
    const { container } = renderWithTheme(<Input caption='Caption' />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('renders input with left icon', () => {
    const { container } = renderWithTheme(
      <Input
        leftIcon={
          <svg width='16' height='16' viewBox='0 0 16 16' fill='none'>
            <path
              fillRule='evenodd'
              clipRule='evenodd'
              d='M8 8C10.21 8 12 6.21 12 4C12 1.79 10.21 0 8 0C5.79 0 4 1.79 4 4C4 6.21 5.79 8 8 8ZM8 10C5.33 10 0 11.34 0 14V15C0 15.55 0.45 16 1 16H15C15.55 16 16 15.55 16 15V14C16 11.34 10.67 10 8 10Z'
              fill='#7F7F7F'
            />
          </svg>
        }
      />
    )
    expect(container.firstChild).toMatchSnapshot()
  })

  it('renders input with right icon', () => {
    const { container } = renderWithTheme(
      <Input
        rightIcon={
          <svg width='16' height='16' viewBox='0 0 16 16' fill='none'>
            <path
              fillRule='evenodd'
              clipRule='evenodd'
              d='M8 8C10.21 8 12 6.21 12 4C12 1.79 10.21 0 8 0C5.79 0 4 1.79 4 4C4 6.21 5.79 8 8 8ZM8 10C5.33 10 0 11.34 0 14V15C0 15.55 0.45 16 1 16H15C15.55 16 16 15.55 16 15V14C16 11.34 10.67 10 8 10Z'
              fill='#7F7F7F'
            />
          </svg>
        }
      />
    )
    expect(container.firstChild).toMatchSnapshot()
  })

  it('renders disabled input', () => {
    const { container } = renderWithTheme(<Input disabled />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('renders error input', () => {
    const { container } = renderWithTheme(<Input error='error message' />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('renders verified input', () => {
    const { container } = renderWithTheme(<Input verified />)
    expect(container.firstChild).toMatchSnapshot()
  })
})
