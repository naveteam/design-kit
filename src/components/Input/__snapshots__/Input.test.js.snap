// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input renders disabled input 1`] = `
.c0 {
  box-sizing: border-box;
  border-radius: 4px;
  -webkit-transition: all 0.3s linear;
  transition: all 0.3s linear;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 328px;
  height: 36px;
  padding: 8px 10px;
  border: 1px solid #7f7f7f;
  border: 1px solid rgba(37,37,37,0.3);
  background: #f5f5f5;
}

.c1 {
  font-family: Roboto;
  font-style: normal;
  font-weight: normal;
  font-size: 16px;
  line-height: 19px;
  border: none;
  outline: none;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  height: 100%;
  background: transparent;
  padding-left: 0;
  padding-right: 0;
  color: #5d5d5d;
}

.c1::-webkit-input-placeholder {
  color: rgba(37,37,37,0.3);
}

.c1::-moz-placeholder {
  color: rgba(37,37,37,0.3);
}

.c1:-ms-input-placeholder {
  color: rgba(37,37,37,0.3);
}

.c1::placeholder {
  color: rgba(37,37,37,0.3);
}

<div
  class=""
  disabled=""
>
  <div
    class="c0"
    disabled=""
  >
    <input
      class="c1"
      disabled=""
    />
  </div>
</div>
`;

exports[`Input renders error input 1`] = `
.c3 {
  font-family: Roboto;
  color: #5d5d5d;
}

.c0 {
  box-sizing: border-box;
  border-radius: 4px;
  -webkit-transition: all 0.3s linear;
  transition: all 0.3s linear;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 328px;
  height: 36px;
  padding: 8px 10px;
  border: 1px solid #7f7f7f;
  border: 1px solid #e53b3b;
}

.c1 {
  font-family: Roboto;
  font-style: normal;
  font-weight: normal;
  font-size: 16px;
  line-height: 19px;
  border: none;
  outline: none;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  height: 100%;
  background: transparent;
  padding-left: 0;
  padding-right: 0;
  color: #5d5d5d;
}

.c1::-webkit-input-placeholder {
  color: rgba(37,37,37,0.3);
}

.c1::-moz-placeholder {
  color: rgba(37,37,37,0.3);
}

.c1:-ms-input-placeholder {
  color: rgba(37,37,37,0.3);
}

.c1::placeholder {
  color: rgba(37,37,37,0.3);
}

.c2 {
  margin: 0;
  margin-top: 5px;
  color: #e53b3b;
  font-size: 12px;
}

<div
  class=""
>
  <div
    class="c0"
  >
    <input
      class="c1"
    />
  </div>
  <p
    class="c2 c3"
  >
    error message
  </p>
</div>
`;

exports[`Input renders input with caption 1`] = `
.c4 {
  font-family: Roboto;
  color: #5d5d5d;
}

.c0:focus-within div {
  border: 1px solid #007dfe;
}

.c0:focus-within p:first-child {
  color: #007dfe;
}

.c1 {
  box-sizing: border-box;
  border-radius: 4px;
  -webkit-transition: all 0.3s linear;
  transition: all 0.3s linear;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 328px;
  height: 36px;
  padding: 8px 10px;
  border: 1px solid #7f7f7f;
}

.c2 {
  font-family: Roboto;
  font-style: normal;
  font-weight: normal;
  font-size: 16px;
  line-height: 19px;
  border: none;
  outline: none;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  height: 100%;
  background: transparent;
  padding-left: 0;
  padding-right: 0;
  color: #5d5d5d;
}

.c2::-webkit-input-placeholder {
  color: rgba(37,37,37,0.3);
}

.c2::-moz-placeholder {
  color: rgba(37,37,37,0.3);
}

.c2:-ms-input-placeholder {
  color: rgba(37,37,37,0.3);
}

.c2::placeholder {
  color: rgba(37,37,37,0.3);
}

.c3 {
  margin: 0;
  margin-top: 5px;
  color: #7f7f7f;
  font-size: 12px;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <input
      class="c2"
    />
  </div>
  <p
    class="c3 c4"
  >
    Caption
  </p>
</div>
`;

exports[`Input renders input with label 1`] = `
.c2 {
  font-family: Roboto;
  color: #5d5d5d;
}

.c0:focus-within div {
  border: 1px solid #007dfe;
}

.c0:focus-within p:first-child {
  color: #007dfe;
}

.c3 {
  box-sizing: border-box;
  border-radius: 4px;
  -webkit-transition: all 0.3s linear;
  transition: all 0.3s linear;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 328px;
  height: 36px;
  padding: 8px 10px;
  border: 1px solid #7f7f7f;
}

.c1 {
  margin: 0;
  margin-bottom: 5px;
  -webkit-transition: all 0.3s linear;
  transition: all 0.3s linear;
}

.c4 {
  font-family: Roboto;
  font-style: normal;
  font-weight: normal;
  font-size: 16px;
  line-height: 19px;
  border: none;
  outline: none;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  height: 100%;
  background: transparent;
  padding-left: 0;
  padding-right: 0;
  color: #5d5d5d;
}

.c4::-webkit-input-placeholder {
  color: rgba(37,37,37,0.3);
}

.c4::-moz-placeholder {
  color: rgba(37,37,37,0.3);
}

.c4:-ms-input-placeholder {
  color: rgba(37,37,37,0.3);
}

.c4::placeholder {
  color: rgba(37,37,37,0.3);
}

<div
  class="c0"
>
  <p
    class="c1 c2"
  >
    Label
  </p>
  <div
    class="c3"
  >
    <input
      class="c4"
    />
  </div>
</div>
`;

exports[`Input renders input with left icon 1`] = `
.c0:focus-within div {
  border: 1px solid #007dfe;
}

.c0:focus-within p:first-child {
  color: #007dfe;
}

.c1 {
  box-sizing: border-box;
  border-radius: 4px;
  -webkit-transition: all 0.3s linear;
  transition: all 0.3s linear;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 328px;
  height: 36px;
  padding: 8px 10px;
  border: 1px solid #7f7f7f;
}

.c2 {
  font-family: Roboto;
  font-style: normal;
  font-weight: normal;
  font-size: 16px;
  line-height: 19px;
  border: none;
  outline: none;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  height: 100%;
  background: transparent;
  padding-left: 4px;
  padding-right: 0;
  color: #5d5d5d;
}

.c2::-webkit-input-placeholder {
  color: rgba(37,37,37,0.3);
}

.c2::-moz-placeholder {
  color: rgba(37,37,37,0.3);
}

.c2:-ms-input-placeholder {
  color: rgba(37,37,37,0.3);
}

.c2::placeholder {
  color: rgba(37,37,37,0.3);
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <svg
      fill="none"
      height="16"
      viewBox="0 0 16 16"
      width="16"
    >
      <path
        clip-rule="evenodd"
        d="M8 8C10.21 8 12 6.21 12 4C12 1.79 10.21 0 8 0C5.79 0 4 1.79 4 4C4 6.21 5.79 8 8 8ZM8 10C5.33 10 0 11.34 0 14V15C0 15.55 0.45 16 1 16H15C15.55 16 16 15.55 16 15V14C16 11.34 10.67 10 8 10Z"
        fill="#7F7F7F"
        fill-rule="evenodd"
      />
    </svg>
    <input
      class="c2"
    />
  </div>
</div>
`;

exports[`Input renders input with right icon 1`] = `
.c0:focus-within div {
  border: 1px solid #007dfe;
}

.c0:focus-within p:first-child {
  color: #007dfe;
}

.c1 {
  box-sizing: border-box;
  border-radius: 4px;
  -webkit-transition: all 0.3s linear;
  transition: all 0.3s linear;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 328px;
  height: 36px;
  padding: 8px 10px;
  border: 1px solid #7f7f7f;
}

.c2 {
  font-family: Roboto;
  font-style: normal;
  font-weight: normal;
  font-size: 16px;
  line-height: 19px;
  border: none;
  outline: none;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  height: 100%;
  background: transparent;
  padding-left: 0;
  padding-right: 4px;
  color: #5d5d5d;
}

.c2::-webkit-input-placeholder {
  color: rgba(37,37,37,0.3);
}

.c2::-moz-placeholder {
  color: rgba(37,37,37,0.3);
}

.c2:-ms-input-placeholder {
  color: rgba(37,37,37,0.3);
}

.c2::placeholder {
  color: rgba(37,37,37,0.3);
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <input
      class="c2"
    />
    <svg
      fill="none"
      height="16"
      viewBox="0 0 16 16"
      width="16"
    >
      <path
        clip-rule="evenodd"
        d="M8 8C10.21 8 12 6.21 12 4C12 1.79 10.21 0 8 0C5.79 0 4 1.79 4 4C4 6.21 5.79 8 8 8ZM8 10C5.33 10 0 11.34 0 14V15C0 15.55 0.45 16 1 16H15C15.55 16 16 15.55 16 15V14C16 11.34 10.67 10 8 10Z"
        fill="#7F7F7F"
        fill-rule="evenodd"
      />
    </svg>
  </div>
</div>
`;

exports[`Input renders verified input 1`] = `
.c0 {
  box-sizing: border-box;
  border-radius: 4px;
  -webkit-transition: all 0.3s linear;
  transition: all 0.3s linear;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 328px;
  height: 36px;
  padding: 8px 10px;
  border: 1px solid #7f7f7f;
  border: 1px solid #17ae29;
}

.c1 {
  font-family: Roboto;
  font-style: normal;
  font-weight: normal;
  font-size: 16px;
  line-height: 19px;
  border: none;
  outline: none;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  height: 100%;
  background: transparent;
  padding-left: 0;
  padding-right: 0;
  color: #5d5d5d;
}

.c1::-webkit-input-placeholder {
  color: rgba(37,37,37,0.3);
}

.c1::-moz-placeholder {
  color: rgba(37,37,37,0.3);
}

.c1:-ms-input-placeholder {
  color: rgba(37,37,37,0.3);
}

.c1::placeholder {
  color: rgba(37,37,37,0.3);
}

<div
  class=""
>
  <div
    class="c0"
  >
    <input
      class="c1"
    />
  </div>
</div>
`;
